google/cloud/pubsub/subscriber_test.cc

```diff
@@ -76,6 +92,103 @@ TEST(SubscriberTest, SubscribeWithOptions) {
+      subscriber.Subscribe([](Message const&, ExactlyOnceAckHandler const&) {})
+          .get());
+}
+
+TEST(SubscriberTest, OptionsClientOverrides) {
+  Subscription const subscription("test-project", "test-subscription");
+  auto mock = std::make_shared<pubsub_mocks::MockSubscriberConnection>();
+  EXPECT_CALL(*mock, options)
+      .WillRepeatedly(Return(Options{}
+                                 .set<TestOptionA>("test-a")
+                                 .set<TestOptionB>("test-b")
+                                 .set<TestOptionC>("test-c")));
+  EXPECT_CALL(*mock, Subscribe).WillOnce([](auto const&) {
+    auto const& current = google::cloud::internal::CurrentOptions();
+    EXPECT_EQ(current.get<TestOptionA>(), "override-a");
+    EXPECT_EQ(current.get<TestOptionB>(), "test-b");
+    EXPECT_EQ(current.get<TestOptionC>(), "test-c");
+    return make_ready_future(Status{});
+  });
+  EXPECT_CALL(*mock, ExactlyOnceSubscribe).WillOnce([](auto const&) {
+    auto const& current = google::cloud::internal::CurrentOptions();
+    EXPECT_EQ(current.get<TestOptionA>(), "override-a");
+    EXPECT_EQ(current.get<TestOptionB>(), "test-b");
+    EXPECT_EQ(current.get<TestOptionC>(), "test-c");
+    return make_ready_future(Status{});
+  });
+
+  Subscriber subscriber(mock, Options{}.set<TestOptionA>("override-a"));
+  ASSERT_STATUS_OK(
+      subscriber.Subscribe([](Message const&, AckHandler const&) {}).get());
+  ASSERT_STATUS_OK(
+      subscriber.Subscribe([](Message const&, ExactlyOnceAckHandler const&) {})
+          .get());
+}
+
+TEST(SubscriberTest, OptionsFunctionOverrides) {
+  Subscription const subscription("test-project", "test-subscription");
+  auto mock = std::make_shared<pubsub_mocks::MockSubscriberConnection>();
+  EXPECT_CALL(*mock, options)
+      .WillRepeatedly(Return(Options{}
+                                 .set<TestOptionA>("test-a")
+                                 .set<TestOptionB>("test-b")
+                                 .set<TestOptionC>("test-c")));
+  EXPECT_CALL(*mock, Subscribe).WillOnce([](auto const&) {
+    auto const& current = google::cloud::internal::CurrentOptions();
+    EXPECT_EQ(current.get<TestOptionA>(), "override-a");
+    EXPECT_EQ(current.get<TestOptionB>(), "override-b1");
+    EXPECT_EQ(current.get<TestOptionC>(), "test-c");
+    return make_ready_future(Status{});
+  });
+  EXPECT_CALL(*mock, ExactlyOnceSubscribe).WillOnce([](auto const&) {
+    auto const& current = google::cloud::internal::CurrentOptions();
+    EXPECT_EQ(current.get<TestOptionA>(), "override-a");
+    EXPECT_EQ(current.get<TestOptionB>(), "override-b2");
+    EXPECT_EQ(current.get<TestOptionC>(), "test-c");
+    return make_ready_future(Status{});
+  });
+
+  Subscriber subscriber(mock, Options{}.set<TestOptionA>("override-a"));
+  ASSERT_STATUS_OK(subscriber
+                       .Subscribe([](Message const&, AckHandler const&) {},
+                                  Options{}.set<TestOptionB>("override-b1"))
```

Comment:
Perhaps these should also be setting `TestOptionA` too, to check that call overrides client.

Reply:
Done